# Processing Text Data
In this Notebook, we will be processing text data using the [Twitter Customer Support](https://www.kaggle.com/datasets/thoughtvector/customer-support-on-twitter). 

Our Notebooks in CSMODEL are designed to be guided learning activities. To use them, simply go through the cells from top to bottom, following the directions along the way. If you find any unclear parts or mistakes in the Notebooks, email your instructor.

## Instructions
* Read each cell and implement the TODOs sequentially. The markdown/text cells also contain instructions which you need to follow to get the whole notebook working.
* Do not change the variable names unless the instructor allows you to.
* You are expected to search how to some functions work on the Internet or via the docs. 
* The notebooks will undergo a 'Restart and Run All' command, so make sure that your code is working properly.
* You are expected to understand the dataset loading and processing separately from this class.
* You may not reproduce this notebook or share them to anyone.

## Import
Import **numpy**, **pandas**, **re**, **nltk**, and **string**.

In [1]:
import numpy as np
import pandas as pd
import re
import nltk
import string
pd.options.mode.chained_assignment = None

%load_ext autoreload
%autoreload 2

## Twitter Customer Support Dataset
For this notebook, we will work on a reduced version of the dataset called `Twitter Customer Support`. The original dataset contains more than 2M rows. In our reduced version, we only retained the first 50k rows in the dataset.

The dataset is provided to you as a `.csv` file. `.csv` means comma-separated values. You can open the file in Notepad to see how it is exactly formatted.

If you view the `.csv` file in Excel, you can see that our dataset contains 50k **observations** (rows) across 7 **variables** (columns). The following are the descriptions of each variable in the dataset.

- **`tweet_id`**: A unique, anonymized ID for the Tweet. Referenced by `response_tweet_id` and `in_response_to_tweet_id`.
- **`author_id`**: A unique, anonymized user ID. @s in the dataset have been replaced with their associated anonymized user ID.
- **`inbound`**: Whether the tweet is "inbound" to a company doing customer support on Twitter. This feature is useful when re-organizing data for training conversational models.
- **`created_at`**: Date and time when the tweet was sent.
- **`text`**: Tweet content. Sensitive information like phone numbers and email addresses are replaced with mask values like `__email__`.
- **`response_tweet_id`**: IDs of tweets that are responses to this tweet, comma-separated.
- **`in_response_to_tweet_id`**: ID of the tweet this tweet is in response to, if any.

Let's read the dataset.

In [2]:
twcs_df = pd.read_csv('twcs-reduced.csv')

Show the first few rows of the dataset.

In [3]:
twcs_df.head()

tweet_id   author_id  inbound                      created_at  \
0         1  sprintcare    False  Tue Oct 31 22:10:47 +0000 2017   
1         2      115712     True  Tue Oct 31 22:11:45 +0000 2017   
2         3      115712     True  Tue Oct 31 22:08:27 +0000 2017   
3         4  sprintcare    False  Tue Oct 31 21:54:49 +0000 2017   
4         5      115712     True  Tue Oct 31 21:49:35 +0000 2017   

                                                text response_tweet_id  \
0  @115712 I understand. I would like to assist y...                 2   
1      @sprintcare and how do you propose we do that               NaN   
2  @sprintcare I have sent several private messag...                 1   
3  @115712 Please send us a Private Message so th...                 3   
4                                 @sprintcare I did.                 4   

   in_response_to_tweet_id  
0                      3.0  
1                      1.0  
2                      4.0  
3                      5.0  
4                      6.0

For this notebook, we will use the values under the `text` column. Thus, let's instantiate a new `DataFrame` with only the `text` column.

In [4]:
twcs_text_df = twcs_df[['text']]
twcs_text_df['text'] = twcs_text_df['text'].astype(str)

Display the first few rows of `twcs_text_df`.

In [5]:
twcs_text_df.head()

text
0  @115712 I understand. I would like to assist y...
1      @sprintcare and how do you propose we do that
2  @sprintcare I have sent several private messag...
3  @115712 Please send us a Private Message so th...
4                                 @sprintcare I did.

## Pre-Processing Text Data

In any machine learning task, preprocessing the data is as important as model building. This process is even more important for unstructured data like texts. In this notebook, we will be performing some of the most common text pre-processing steps including:

* Lower casing
* Removal of Punctuations
* Removal of Stopwords
* Removal of Frequent words
* Stemming
* Lemmatization


Do note that all of these pre-processing steps need not be performed on the dataset all the time. You need to carefully identify appropriate pre-processing techniques depending on the data or the task. For example, emojis or emoticons might be useful in sentiment analysis, thus it might not be a good idea to remove them.

Open `text_preprocessor.py` file. Some of the functions in the file are not yet implemented. We will implement the missing functions of this file.

### Lower Casing

Lower casing converts the input text into the same case so that 'text', 'Text', and 'TEXT' are similarly treated. This is especially helpful in getting the correct frequency of the same word but are represented in different cases. However, this may not be helpful when performing Part-of-Speech tagging (where proper casing gives some information about Nouns and so on) or sentiment analysis (where upper casing refers to anger and so on).

By default, lower casing is done my most of the modern day vectorizers and tokenizers like [sklearn TfidfVectorizer](https://scikit-learn.org/stable/modules/generated/sklearn.feature_extraction.text.TfidfVectorizer.html). Thus, set them to `False` as needed depending on the use case. 

Open `text_preprocessor.py` file and complete the `to_lower_case()` function. This converts characters in each string of a Series to lower case.

Implement the `to_lower_case()` function. Inline comments should help you in completing the contents of the function.

Afterwards, let's import the function.

In [6]:
# Write your code here
import text_preprocessor as tp

Convert the texts in column `text` to lowercase by calling the function `to_lower_case()` and assign the return value to a new column `text_lower`.

In [7]:
# Write your code here
text_lower = tp.to_lower_case(twcs_text_df['text'])

Let's display the `DataFrame`.

In [8]:
twcs_text_df.head()

text
0  @115712 I understand. I would like to assist y...
1      @sprintcare and how do you propose we do that
2  @sprintcare I have sent several private messag...
3  @115712 Please send us a Private Message so th...
4                                 @sprintcare I did.

Display the lowercase version of the string in index `10`.

In [9]:
# Write your code here
text_lower.loc[10]

'@sprintcare since i signed up with you....since day 1'

**Checkpoint #1:** After calling the function `to_lower_case()`, what is the string in index `10`?

Answer: @sprintcare since i signed up with you....since day 1

Display the lowercase version of the string in index `100`.

In [10]:
# Write your code here
text_lower.loc[100]

"@askplaystation so, what's the november ps plus free game?"

**Checkpoint #2:** After calling the function `to_lower_case()`, what is the string in index `100`?

Answer: @askplaystation so, what's the november ps plus free game?

### Removal of Punctuations

Removing punctuations also standardizes the text data. This will treat the words 'hurray!' and 'hurray' in the same way. The list of punctuations to include should depend on the use case. For example, the `string.punctuation` in python contains the following punctuation symbols:

``!"#$%&\'()*+,-./:;<=>?@[\\]^_`{|}~``

Add or remove punctuations depending on the use case.

Get the punctuations in `string.punctuation`.

In [11]:
punctuations = string.punctuation
print(punctuations)

!"#$%&'()*+,-./:;<=>?@[\]^_`{|}~


Open `text_preprocessor.py` file and complete the `remove_punctuations()` function. This takes in a `Series` of strings and removes punctuations in the text.

Implement the `remove_punctuations()` function. Inline comments should help you in completing the contents of the function.

Afterwards, let's import the function.

In [12]:
# Write your code here
from text_preprocessor import remove_punctuations

Remove punctuations in the texts in column `text_lower` by calling the function `remove_punctuations()` and assign the return value to a new column `text_wo_punct`.

In [13]:
# Write your code here
text_wo_punct = tp.remove_punctuations(text_lower)

Let's display the `DataFrame`.

In [14]:
twcs_text_df.head()

text
0  @115712 I understand. I would like to assist y...
1      @sprintcare and how do you propose we do that
2  @sprintcare I have sent several private messag...
3  @115712 Please send us a Private Message so th...
4                                 @sprintcare I did.

Display the string without punctuations in index `300`.

In [15]:
# Write your code here
text_wo_punct[300]

'mcdonalds treat'

**Checkpoint #3:** After calling the function `remove_punctuations()`, what is the string in index `300`?

Answer: mcdonalds treat

Display the string without punctuations in index `1000`.

In [16]:
# Write your code here
text_wo_punct[1000]

'bofahelp whenever your ready'

**Checkpoint #4:** After calling the function `remove_punctuations()`, what is the string in index `1000`?

Answer: bofahelp whenever your ready

### Removal of Stopwords

Stopwords are commonly occuring words in a language, which include 'the', 'a', among others. These words can be removed from the text most of the time since they do not provide valuable information for analysis. However, these words might be important when performing Part-of-Speech tagging.

List of stopwords are already compiled for different languages. For example, the list of stopwords for the English language from the `nltk` package can be seen below.

Let's download `stopwords` from the `nltk.corpus` package.

In [17]:
nltk.download('stopwords')

[nltk_data] Downloading package stopwords to
[nltk_data]     /Users/angeloguerra/nltk_data...
[nltk_data]   Package stopwords is already up-to-date!


True

Import `stopwords` and print the stopwords in English.

In [18]:
from nltk.corpus import stopwords
print(stopwords.words('english'))

['i', 'me', 'my', 'myself', 'we', 'our', 'ours', 'ourselves', 'you', "you're", "you've", "you'll", "you'd", 'your', 'yours', 'yourself', 'yourselves', 'he', 'him', 'his', 'himself', 'she', "she's", 'her', 'hers', 'herself', 'it', "it's", 'its', 'itself', 'they', 'them', 'their', 'theirs', 'themselves', 'what', 'which', 'who', 'whom', 'this', 'that', "that'll", 'these', 'those', 'am', 'is', 'are', 'was', 'were', 'be', 'been', 'being', 'have', 'has', 'had', 'having', 'do', 'does', 'did', 'doing', 'a', 'an', 'the', 'and', 'but', 'if', 'or', 'because', 'as', 'until', 'while', 'of', 'at', 'by', 'for', 'with', 'about', 'against', 'between', 'into', 'through', 'during', 'before', 'after', 'above', 'below', 'to', 'from', 'up', 'down', 'in', 'out', 'on', 'off', 'over', 'under', 'again', 'further', 'then', 'once', 'here', 'there', 'when', 'where', 'why', 'how', 'all', 'any', 'both', 'each', 'few', 'more', 'most', 'other', 'some', 'such', 'no', 'nor', 'not', 'only', 'own', 'same', 'so', 'than', '

Open `text_preprocessor.py` file and complete the `remove_stopwords()` function. This takes in a `Series` of strings and removes stopwords in the text.

Implement the `remove_stopwords()` function. Inline comments should help you in completing the contents of the function.

Afterwards, let's import the function.

In [19]:
# Write your code here
from text_preprocessor import remove_stopwords

Remove punctuations in the texts in column `text_wo_punct` by calling the function `remove_stopwords()` and assign the return value to a new column `text_wo_stop`.

In [20]:
# Write your code here
text_wo_stop = tp.remove_stopwords(text_wo_punct )

Let's display the `DataFrame`.

In [21]:
twcs_text_df.head()

text
0  @115712 I understand. I would like to assist y...
1      @sprintcare and how do you propose we do that
2  @sprintcare I have sent several private messag...
3  @115712 Please send us a Private Message so th...
4                                 @sprintcare I did.

Display the string without stopwords in index `18`.

In [22]:
# Write your code here
text_wo_stop[18]

'askspectrum received corporate office would like copy'

**Checkpoint #5:** After calling the function `remove_stopwords()`, what is the string in index `18`?

Answer: askspectrum received corporate office would like copy

Display the string without stopwords in index `3000`.

In [23]:
# Write your code here
text_wo_stop[3000]

'britishairways satisfactorily dm please'

**Checkpoint #6:** After calling the function `remove_stopwords()`, what is the string in index `3000`?

Answer: britishairways satisfactorily dm please

### Removal of Frequent Words

If you are working on a domain-specific corpus, most of the frequent words might not be important in processing the text data. Thus, it might be useful to remove frequent words in the given corpus. If you a technique similar to tf-idf, this is automatically taken care of.

Open `text_preprocessor.py` file and complete the `get_frequent_words()` function. This returns the most frequent words in our dataset.

Implement the `get_frequent_words()` function. Inline comments should help you in completing the contents of the function.

Afterwards, let's import the function.

In [24]:
# Write your code here
from text_preprocessor import get_frequent_words

Let's call the function to get the top 15 most frequent words in the texts in column `text_wo_stop`. Display the words.

In [25]:
# Write your code here
tp.get_frequent_words(text_wo_stop, 15)

[('us', 7208),
 ('please', 6590),
 ('dm', 4998),
 ('help', 4376),
 ('hi', 3758),
 ('thanks', 3594),
 ('get', 3350),
 ('sorry', 3036),
 ('amazonhelp', 2479),
 ('know', 2451),
 ('like', 2263),
 ('look', 2159),
 ('well', 2146),
 ('send', 2127),
 ('im', 2069)]

**Checkpoint #7:** What is the most frequent word in the dataset?

Answer: ('us', 7208)

**Checkpoint #8:** What is the 5th most frequent word in the dataset?

Answer: ('hi', 3758)

**Checkpoint #9:** What is the 10th most frequent word in the dataset?

Answer: ('know', 2451)

**Checkpoint #10:** What is the 15th most frequent word in the dataset?

Answer: ('im', 2069)

Open `text_preprocessor.py` file and complete the `remove_frequent_words()` function. This takes in a `Series` of strings and removes the top frequent words in the dataset.

Implement the `remove_frequent_words()` function. Inline comments should help you in completing the contents of the function.

Afterwards, let's import the function.

In [26]:
# Write your code here
from text_preprocessor import remove_frequent_words

Remove the top 10 frequent words in the texts in column `text_wo_stop` by calling the function `remove_frequent_words()` and assign the return value to a new column `text_wo_freq`.

In [27]:
# Write your code here
text_wo_freq = tp.remove_frequent_words(text_wo_stop, 10)

Let's display the `DataFrame`.

In [28]:
twcs_text_df.head()

text
0  @115712 I understand. I would like to assist y...
1      @sprintcare and how do you propose we do that
2  @sprintcare I have sent several private messag...
3  @115712 Please send us a Private Message so th...
4                                 @sprintcare I did.

Display the string without frequent words in index `43687`.

In [29]:
# Write your code here
text_wo_freq[43687]

'tesco peter getting back'

**Checkpoint #11:** After calling the function `remove_frequent_words()`, what is the string in index `43687`?

Answer: tesco peter getting back

Display the string without frequent words in index `44762`.

In [30]:
# Write your code here
text_wo_freq[44762]

'askpaypal still needing sent'

**Checkpoint #12:** After calling the function `remove_frequent_words()`, what is the string in index `44762`?

Answer: askpaypal still needing sent

### Stemming

Stemming is the process of reducing inflected (or sometimes derived) words to their word stem, base or root form (From [Wikipedia](https://en.wikipedia.org/wiki/Stemming)).

For example, stemming will remove the suffixes in words like 'walks' and 'walking' to convert them to the root word 'walk'. But in another example, we have two words 'console' and 'consoling', the stemmer will remove the suffix and make them 'consol' which is not a proper english word.

Open `text_preprocessor.py` file and complete the `stem()` function. This takes in a `Series` of strings and performs stemming to each word in the string. This uses the `PorterStemmer` from the `nltk` package.

Implement the `stem()` function. Inline comments should help you in completing the contents of the function.

Import the function.

In [31]:
# Write your code here
from text_preprocessor import stem

Perform stemming in the texts in column `text_wo_freq` by calling the function `stem()` and assign the return value to a new column `text_stem`.

In [32]:
# Write your code here
text_stem = tp.stem(text_wo_freq)

Let's display the `DataFrame`.

In [33]:
twcs_text_df.head()

text
0  @115712 I understand. I would like to assist y...
1      @sprintcare and how do you propose we do that
2  @sprintcare I have sent several private messag...
3  @115712 Please send us a Private Message so th...
4                                 @sprintcare I did.

Words like 'private' and 'propose' have their 'e' at the end chopped off due to stemming. This is not intented. Lemmatization is used in such cases.

Display the string with stemmed words in index `18`.

In [34]:
# Write your code here
text_stem[18]

'askspectrum receiv corpor offic would like copi'

**Checkpoint #13:** After calling the function `stem()`, what is the string in index `18`?

Answer: askspectrum receiv corpor offic would like copi

Display the string with stemmed words in index `100`.

In [35]:
# Write your code here
text_stem[100]

'askplayst what novemb ps plu free game'

**Checkpoint #14:** After calling the function `stem()`, what is the string in index `100`?

Answer: askplayst what novemb ps plu free game

### Lemmatization

Lemmatization is similar to stemming in reducing inflected words to their word stem, but differs in the way that it makes sure the root word (also called as lemma) belongs to the language. As a result, this is generally slower than stemming process. Thus, either stemming or lemmatization can be used depending on the speed requirement.

Let's download `wordnet` from the `nltk` package. This is needed for lemmatization.

In [36]:
nltk.download('wordnet')

[nltk_data] Downloading package wordnet to
[nltk_data]     /Users/angeloguerra/nltk_data...
[nltk_data]   Package wordnet is already up-to-date!


True

We will use the `WordNetLemmatizer` from the `nltk` package.

In [37]:
from nltk.stem import WordNetLemmatizer
lemmatizer = WordNetLemmatizer()

Let's try to perform lemmatization on the word 'running'. This should return the root word 'run'.

In [38]:
lemmatizer.lemmatize('running')

'running'

Notice that the lemmatizer returns the word 'running' instead of the root word 'run'. This is because the lemmatization process depends on the Part-of-Speech tag to come up with the correct lemma. 

Lemmatize the word 'running' again by providing the correct Part-of-Speech tag `v` for verbs.

In [39]:
lemmatizer.lemmatize('running', 'v')

'run'

Now we are getting the root form 'run'. Thus, there is a need to provide the Part-of-Speech tag of the word along with the word for lemmatizer in `nltk`. Depending on the tag, the lemmatizer may return different results.

Perform lemmatization on the word 'stripes' and check the lemma when it is both verb and noun.

In [40]:
print('Lemma result for verb : ', lemmatizer.lemmatize('stripes', 'v'))
print('Lemma result for noun : ', lemmatizer.lemmatize('stripes', 'n'))

Lemma result for verb :  strip
Lemma result for noun :  stripe


Open `text_preprocessor.py` file and complete the `lemmatize()` function. This takes in a `Series` of strings and performs lemmatization to each word in the string. This uses the `WordNetLemmatizer` and the `pos_tag` from the `nltk` package.

Implement the `lemmatize()` function. Inline comments should help you in completing the contents of the function.

Import the function.

In [41]:
# Write your code here
from text_preprocessor import lemmatize

Let's download `averaged_perceptron_tagger` from the `nltk` package. We will use this to get the Part-of-Speech tag of each word in a string.

In [42]:
nltk.download('averaged_perceptron_tagger')

[nltk_data] Downloading package averaged_perceptron_tagger to
[nltk_data]     /Users/angeloguerra/nltk_data...
[nltk_data]   Package averaged_perceptron_tagger is already up-to-
[nltk_data]       date!


True

Perform lemmatization in the texts in column `text_wo_freq` by calling the function `lemmatize()` and assign the return value to a new column `text_lemma`.

In [43]:
# Write your code here
text_lemma = tp.lemmatize(text_wo_freq)

Let's display the `DataFrame`.

In [44]:
twcs_text_df.head()

text
0  @115712 I understand. I would like to assist y...
1      @sprintcare and how do you propose we do that
2  @sprintcare I have sent several private messag...
3  @115712 Please send us a Private Message so th...
4                                 @sprintcare I did.

We can see that the trailing 'e' in the words 'propose' and 'private' is retained when we use lemmatization unlike stemming. 

Display the string with lemmatized words in index `49267`.

In [45]:
# Write your code here
text_lemma[49267]

'virginatlantic propose resolve'

**Checkpoint #15:** After calling the function `lemmatize()`, what is the string in index `49267`?

Answer: virginatlantic propose resolve

Display the string with lemmatized words in index `750`.

In [46]:
# Write your code here
text_lemma[750]

'sprintcare ive send several private message one responds'

**Checkpoint #16:** After calling the function `lemmatize()`, what is the string in index `750`?

Answer: sprintcare ive send several private message one responds